src/mmrm_spatial.cpp

```diff
@@ -0,0 +1,134 @@
+//
+// Note: This is a special generalized least squares model
+// Y = X * beta + epsilon,
+// where we have a block structure for the covariance matrix of the epsilon
+// vector.
+//
+// beta itself is not a parameter for TMB here:
+// - For maximum likelihood estimation:
+//   Given theta and therefore Sigma, and writing W = Sigma^-1, we can determine
+//   the beta optimizing the likelihood via the weighted least squares equation
+//   (X^T W X) beta = X^T W Y.
+// - For restricted maximum likelihood estimation:
+//   Given theta, beta is integrated out from the likelihood. Weighted least
+//   squares results are used to calculate integrated log likelihood.
+
+template<class Type>
+Type objective_function<Type>::operator() ()
+{
+  // Read data from R.
+  DATA_MATRIX(x_matrix);           // Model matrix (dimension n x p).
+  DATA_VECTOR(y_vector);           // Response vector (length n).
+  DATA_SPARSE_MATRIX(distance);    // Coordinates matrix (block diagonal sparse matrix).
+  DATA_INTEGER(n_subjects);        // Number of subjects.
+  DATA_IVECTOR(subject_zero_inds); // Starting indices for each subject (0-based) (length n_subjects).
+  DATA_IVECTOR(subject_n_visits);  // Number of observed visits for each subject (length n_subjects).
+  DATA_STRING(cov_type);           // Covariance type name.
+  DATA_INTEGER(reml);              // REML (1)? Otherwise ML (0).
+  DATA_FACTOR(subject_groups);     // subject groups vector(0-based) (length n_subjects).
+  DATA_INTEGER(n_groups);          // number of total groups.
+  // Read parameters from R.
+  PARAMETER_VECTOR(theta);         // Covariance parameters (length k). Contents depend on covariance type.
+
+  // X^T W X will be calculated incrementally into here.
+  matrix<Type> XtWX = matrix<Type>::Zero(x_matrix.cols(), x_matrix.cols());
+  // X^T W Y will be calculated incrementally into here.
+  matrix<Type> XtWY = matrix<Type>::Zero(x_matrix.cols(), 1);
+  // W^T/2 X will be saved into here.
+  matrix<Type> x_mat_tilde = matrix<Type>::Zero(x_matrix.rows(), x_matrix.cols());
+  // W^T/2 Y will be saved into here.
+  vector<Type> y_vec_tilde = vector<Type>::Zero(y_vector.rows());
+  // Sum of the log determinant will be incrementally calculated here.
+  Type sum_log_det = 0.0;
+  // Create the lower triangular Cholesky factor of the visit x visit covariance matrix.
+  int theta_size = theta.size() / n_groups;
+  // Go through all subjects and calculate quantities initialized above.
+  for (int i = 0; i < n_subjects; i++) {
+    // Start index and number of visits for this subject.
+    int start_i = subject_zero_inds(i);
+    int n_visits_i = subject_n_visits(i);
+    // Create the distance matrix.
+    matrix<Type> distance_i = distance.block(start_i, start_i, n_visits_i, n_visits_i);
+    // Obtain Cholesky factor Li.
+    vector<Type> theta_i = theta.segment(subject_groups(i) * theta_size, theta_size);
+    matrix<Type> cov_i = get_spatial_covariance_lower_chol(theta_i, distance_i);
```

Comment:
This should be the only structural difference from a non "spatial" (still don't like the term) covariance structure, right? Besides, passing in the precalculated distances. There might even be a case to be made for calculating the distances within TMB to keep the changes to the rest of the API small. I doubt that comes at a large performance hit.